### PR TITLE
Add performance testing setup (close #37)

### LINKED ISSUE #37: Add performance testing setup
In order to make sure that future changes don't come at the cost of performance, we need to monitor performance. A common framework for measuring performance should be set up so that it will enable performance comparison.

### PR DESCRIPTION
**Pull request by matus-tomlein**
Issue: #37 

## Overview

This PR adds a new executable into the project that runs simple performance tests and logs their output. The goal is to be able to measure effect of changes in the tracker code on performance.

The tests only measure performance of tracking events without sending events over network. The idea is to measure the impact of the tracker on the "main" thread or the thread where events are tracked. The performance tests do not account for asynchronous operations (such as sending events).

The aim is to evaluate performance by comparison. There are several comparison that the tests enable:

1. Comparison **with past performance**. Output of each run is logged into a `performance/logs.txt` file and there is a script that extracts statistics from the historical and last performance.
2. Comparison of performance **with and without emitter and session**. We may add other variations in the future, but currently I think that the emitter and session have the biggest impact on performance. The tests compare the performance in 4 variations:
    1. mocked emitter and mocked session
    2. mocked emitter and real session
    3. mute emitter (standard with saving to DB, but without sending background loop) and mocked session
    4. mute emitter and real session

Each variation runs the same procedure – it tracks 10 000 timing, 10 000 screen view, 10 000 structured events in 5 threads in parallel (i.e., 150 000 events).

## Running the performance tests

To run performance tests under macOS, you will need to build the project using `make` and run the tests:

```bash
 host> ./build/performance/tracker_performance
```

## Performance logs

Each run of the tests appends a JSON to the `performance/logs.txt` file with desktop context describing the current machine and statistics extracted from the test. One measurement looks like this:

```json
{
    "desktop_context": {
        "data": {
            "deviceManufacturer": "Apple Inc.",
            "deviceModel": "MacBookPro17,1",
            "deviceProcessorCount": 8,
            "osIs64Bit": true,
            "osServicePack": "",
            "osType": "macOS",
            "osVersion": "12.2.0"
        },
        "schema": "iglu:com.snowplowanalytics.snowplow/desktop_context/jsonschema/1-0-0"
    },
    "results": {
        "mocked_emitter_and_mocked_session": 5.273350042,
        "mocked_emitter_and_real_session": 5.072687333,
        "mute_emitter_and_mocked_session": 18.7668315,
        "mute_emitter_and_real_session": 22.612970792,
        "num_operations": 10000,
        "num_threads": 5
    },
    "timestamp": 1644589599181,
    "tracker_version": "cpp-0.1.0"
}

```

## Comparing with historical performance

There is a simple Python script that compares the last run with historical performance. It outputs a table with several statistics.

To run the comparison script:

```bash
 host> ./performance/stats.py
```

This will generate a table something like this:

| Metric                                | Max     | Min     | Mean    | Last    |
|---|---|---|---|---|
| mocked emitter and mocked session     | 5.27s   | 5.08s   | 5.18s   | 5.27s   |
| mocked emitter and real session       | 5.08s   | 5.04s   | 5.06s   | 5.07s   |
| mute emitter and mocked session       | 18.77s  | 17.29s  | 18.08s  | 18.77s  |
| mute emitter and real session         | 28.02s  | 22.61s  | 25.06s  | 22.61s  |

### CODE REVIEW

**Comment by snowplowcla** [comment]
Thanks for your pull request. Is this your first contribution to a Snowplow open source project? Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).

:memo: Please visit https://docs.snowplowanalytics.com/docs/contributing/contributor-license-agreement/ to learn more and sign.

Once you've signed, please reply here (e.g. I signed it!) and we'll verify. Thanks.

**Comment by matus-tomlein** [comment]
@mscwilson Hey Miranda, I thought I'd tag you here in case this is interesting for you since you were working on similar performance testing tasks. No need to review the C++ code, probably more interesting is the general approach and procedure for performance testing. If you have any suggestions or would like to chat about it, let me know.

File: performance/main.cpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+#include "../src/utils.hpp"
+
+using namespace std;
+
+int main(int argc, char **argv) {
+  // run and measure performance
+  double mocked_emitter_and_mocked_session = run_mocked_emitter_and_mocked_session("demo0.db");
+  double mocked_emitter_and_real_session = run_mocked_emitter_and_real_session("demo1.db");
+  double mute_emitter_and_mocked_session = run_mute_emitter_and_mocked_session("demo2.db");
+  double mute_emitter_and_real_session = run_mute_emitter_and_real_session("demo3.db");
```
Comment (AlexBenny) [inline performance/main.cpp]: Just a curiosity: why do we need different files for the storage?
Is it possible some kind of interference?

**Review by AlexBenny — APPROVED** [review]
Looks great, it seems a great first step into the performance testing!
The script to compare results is also nice.
Curious if you have any suggestion or idea about how to integrate this in our CI workflow.

**Comment by mscwilson** [comment]
For the Python comparison table, can you add a column for `n`? Would give more context to the mean and other values.

File: performance/main.cpp (hunk 1 above)
Comment (matus-tomlein) [reply performance/main.cpp]: That's actually a good point, there was no need to use separate databases as the tests are run in sequence. I changed it to use the same `performance.db` database and clean it after each run.

**Comment by matus-tomlein** [comment]
Thanks for the comments, both!

Miranda, I changed the Python script to also show the number of operations and threads.

Alex, I don't have a clear idea yet how to integrate with CI. It would be ideal to run the tests on every tagged release, but not sure how to store the results of the test. One option would be to add it as asset to the release as Paul suggested yesterday. It requires some more thought, let's first see how useful the tests are as we change stuff in the tracker.

**Pull request merged**